server/session/handler_grindstone.go

```diff
@@ -0,0 +1,111 @@
+package session
+
+import (
+	"fmt"
+	"github.com/df-mc/dragonfly/server/block"
+	"github.com/df-mc/dragonfly/server/entity"
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/go-gl/mathgl/mgl64"
+	"github.com/sandertv/gophertunnel/minecraft/protocol"
+	"math"
+	"math/rand"
+)
+
+const (
+	// grindstoneFirstInputSlot is the slot index of the first input item in the grindstone.
+	grindstoneFirstInputSlot = 0x10
+	// grindstoneSecondInputSlot is the slot index of the second input item in the grindstone.
+	grindstoneSecondInputSlot = 0x11
+)
+
+// handleGrindstoneCraft handles a CraftGrindstoneRecipe stack request action made using a grindstone.
+func (h *ItemStackRequestHandler) handleGrindstoneCraft(s *Session) error {
+	// First check if there actually is a grindstone opened.
+	if !s.containerOpened.Load() {
+		return fmt.Errorf("no grindstone container opened")
+	}
+	if _, ok := s.c.World().Block(s.openedPos.Load()).(block.Grindstone); !ok {
+		return fmt.Errorf("no grindstone container opened")
+	}
+
+	// Next, get both input items and ensure they are comparable.
+	firstInput, _ := h.itemInSlot(protocol.StackRequestSlotInfo{
+		ContainerID: containerGrindstoneFirstInput,
+		Slot:        grindstoneFirstInputSlot,
+	}, s)
+	secondInput, _ := h.itemInSlot(protocol.StackRequestSlotInfo{
+		ContainerID: containerGrindstoneSecondInput,
+		Slot:        grindstoneSecondInputSlot,
+	}, s)
+	if firstInput.Empty() && secondInput.Empty() {
+		return fmt.Errorf("input item(s) are empty")
+	}
+	if firstInput.Count() > 1 || secondInput.Count() > 1 {
+		return fmt.Errorf("input item(s) are not single items")
+	}
+
+	resultStack := existingItem(firstInput, secondInput)
+	if !firstInput.Empty() && !secondInput.Empty() {
+		resultStack = firstInput.WithEnchantments(secondInput.Enchantments()...)
```

Comment:
Is this correct? Isn't the output item always supposed to be enchantment free?

Reply:
Ah I see, this is to get the total experience from all those enchantments later on. Might want to document that.